features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/src/main/java/org/opennms/features/topology/plugins/topo/graphml/GraphMLTopologyProvider.java

```diff
@@ -70,10 +75,28 @@ private static TopologyProviderInfo createTopologyProviderInfo(GraphMLGraph grap
     private final List<String> focusIds;
     private final Boolean requiresStatusProvider;
 
-    public GraphMLTopologyProvider(GraphMLGraph graph) {
+    public GraphMLTopologyProvider(GraphMLGraph graph, GraphMLServiceAccessor serviceAccessor) {
         super(graph.getProperty(GraphMLProperties.NAMESPACE));
+
+        m_serviceAccessor = serviceAccessor;
+
         for (GraphMLNode graphMLNode : graph.getNodes()) {
             GraphMLVertex newVertex = new GraphMLVertex(this.getVertexNamespace(), graphMLNode);
+
+            if (newVertex.getNodeID() == null &&
+                    !Strings.isNullOrEmpty((String) newVertex.getProperties().get(GraphMLProperties.FOREIGN_SOURCE)) &&
+                    !Strings.isNullOrEmpty((String) newVertex.getProperties().get(GraphMLProperties.FOREIGN_ID))
+                    ) {
+
+                OnmsNode onmsNode = m_serviceAccessor.getNodeDao().findByForeignId(
+                        (String) newVertex.getProperties().get(GraphMLProperties.FOREIGN_SOURCE),
+                        (String) newVertex.getProperties().get(GraphMLProperties.FOREIGN_ID));
+
+                if (onmsNode != null) {
+                    newVertex.setNodeID(onmsNode.getId());
+                }
+            }
```

Comment:
We should also add a warning, if either foreignSource or foreignId is set, but not both.